presto-iceberg/src/main/java/io/prestosql/iceberg/IcebergSplitSource.java

```diff
@@ -0,0 +1,212 @@
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static io.prestosql.iceberg.IcebergUtil.getIdentityPartitions;
+import static io.prestosql.plugin.hive.HivePartitionKey.HIVE_DEFAULT_DYNAMIC_PARTITION;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public class IcebergSplitSource
+        implements ConnectorSplitSource
+{
+    public static final String LOCALHOST = "localhost";
+
+    private final String database;
+    private final String tableName;
+    private final Iterator<CombinedScanTask> scanTaskIterator;
+    private final TupleDomain<HiveColumnHandle> predicates;
+    private final ConnectorSession session;
+    private final Schema tableSchema;
+    private final HdfsEnvironment hdfsEnvironment;
+    private final TypeTranslator typeTranslator;
+    private final TypeManager typeRegistry;
+    private final HdfsEnvironment.HdfsContext hdfsContext;
+    private boolean closed;
+    private Map<String, HiveColumnHandle> columnNameToHiveColumnHandleMap;
+    private final Long snapshotId;
+    private final Long snapshotTimestamp;
+
+    public IcebergSplitSource(String database,
+            String tableName,
+            Iterator<CombinedScanTask> fileScanTaskIterator,
+            TupleDomain<HiveColumnHandle> predicates,
+            ConnectorSession session,
+            Schema schema,
+            HdfsEnvironment hdfsEnvironment,
+            TypeTranslator typeTranslator,
+            TypeManager typeRegistry,
+            Map<String, HiveColumnHandle> columnNameToHiveColumnHandleMap,
+            Long snapshotId,
+            Long snapshotTimestamp)
+    {
+        this.database = database;
+        this.tableName = tableName;
+        this.scanTaskIterator = fileScanTaskIterator;
+        this.predicates = predicates;
+        this.session = session;
+        this.tableSchema = schema;
+        this.hdfsEnvironment = hdfsEnvironment;
+        this.snapshotId = snapshotId;
+        this.snapshotTimestamp = snapshotTimestamp;
+        this.hdfsContext = new HdfsEnvironment.HdfsContext(session, database, tableName);
+        this.typeTranslator = typeTranslator;
+        this.typeRegistry = typeRegistry;
+        this.columnNameToHiveColumnHandleMap = columnNameToHiveColumnHandleMap;
+    }
+
+    @Override
+    public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+    {
+        List<ConnectorSplit> splits = new ArrayList<>();
+        final TupleDomain<HiveColumnHandle> predicates = DomainConverter.handleTypeDifference(this.predicates);
+        while (scanTaskIterator.hasNext() && maxSize != 0) {
```

Comment:
If the size of `combinedScanTask.files()` is not a multiple of `maxSize`, `maxSize` can become negative?